File: 9-Stack/programs/ruby/program4.rb
class Node
  attr_accessor :data, :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

class Stack
  def initialize
    @top = nil
  end

  def push(x)
    t = Node.new(x)
    if t.nil?
      puts "Stack is full"
    else
      t.next = @top
      @top = t
    end
  end

  def pop
    x = nil
    if @top.nil?
      puts "Stack is Empty"
    else
      t = @top
      @top = @top.next
      x = t.data
    end
    x
  end

  def display
    p = @top
    while p
      print "#{p.data} "
      p = p.next
    end
    puts
  end

  def peek(pos)
    p = @top
    (pos - 1).times do
      p = p.next if p
    end
    p ? p.data : -1
  end

  def stack_top
    @top ? @top.data : -1
  end

  def is_empty
    @top.nil? ? 1 : 0
  end

  def is_full
    t = Node.new(nil)
    r = t.nil? ? 1 : 0
    t = nil
    r
  end

  def is_balanced(exp)
    exp.each_char do |char|
      if char == '('
        push(char)
      elsif char == ')'
        return 0 if @top.nil?
        pop
      end
    end
    @top.nil? ? 1 : 0
  end
end

exp = "((a+b)*(c-d))"
stack = Stack.new
puts stack.is_balanced(exp)

File: 9-Stack/programs/cpp/program4.cpp
// Stack implementation

#include<bits/stdc++.h>
using namespace std;

struct Node{
  char data;
  struct Node *next;
}*top = NULL;

void push(char x){
  struct Node *t;
  t = new Node;
  if(t == NULL){
    cout << "Stack is full" << endl;
  }else{
    t->data = x;
    t->next = top;
    top = t;
  }
}

char pop(){
  char x = -1;
  if(top == NULL){
    cout << "Stack is Empty" << endl;
  }else{
    struct Node *t = top;
    top = top->next;
    x = t->data;
    delete t;
  }
  return x;
}

void Display(){
  struct Node *p = top;
  while(p != NULL){
    cout << p->data << " ";
    p = p->next;
  }
  cout << endl;
}

int peek(int pos){
  struct Node *p = top;
  for(int i = 0; p != NULL && i < pos-1; i++){
    p = p->next;
  }
  if(p != NULL){
    return p->data;
  }else{
    return -1;
  }
}


int stackTop(){
  if(top != NULL){
    return top->data;
  }else{
    return -1;
  }
}

int isEmpty(){
  return top ? 0 : 1;
}

int isFull(){
  struct Node *t = new Node;
  int r = t ? 0 : 1;
  delete t;
  return r;
}

int isBalanced(char *exp){
  for(int i = 0; exp[i] != '\0'; i++){
    if(exp[i] == '('){
      push(exp[i]);
    }else if(exp[i] == ')'){
      if(top == NULL){
        return 0;
      }
      pop();
    }
  }
  return top == NULL ? 1 : 0;
}


int main(){
 char *exp = "((a+b)*(c-d))))";
 cout << isBalanced(exp) << endl;
  return 0;
}



